### PR TITLE
Feature/tm2i message extras

### PR DESCRIPTION
**Pull request by marto83**
I haven't looked at the java implementation tests so I might need to add some more.

### CODE REVIEW

File: src/IO.Ably.Shared/Types/MessageExtras.cs [hunk 1]
```diff
@@ -1,14 +1,64 @@
...
 {
     /// <summary>
     /// Extra properties on the Message.
     /// </summary>
     public class MessageExtras
     {
+        private const string DeltaProperty = "delta";
+
+        [JsonIgnore]
+        private JToken Data { get; set; }
```
Comment (QuintinWillison) [inline src/IO.Ably.Shared/Types/MessageExtras.cs:14]: Per [the Java implementation](https://github.com/ably/ably-java/pull/581/files), the `MessageExtras` instances should be immutable. I would expect the fields backing `Data` and `Delta` to be `readonly`.

Can you please look at your alignment with the Java implementation.

File: src/IO.Ably.Shared/CustomSerialisers/MessageExtrasConverter.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+using System;
+using IO.Ably.Types;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace IO.Ably.CustomSerialisers
+{
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+#pragma warning disable SA1600 // Elements should be documented
+    public class MessageExtrasConverter : JsonConverter
```
Comment (QuintinWillison) [inline src/IO.Ably.Shared/CustomSerialisers/MessageExtrasConverter.cs:10]: Now would be a good opportunity to add unit tests for classes like this.

If nothing else a 'loop back' test that validates conversion to JSON and then back end yields a new instance the 'equals' the origin instance.

File: src/IO.Ably.Shared/Types/MessageExtras.cs (hunk 1 above)
Comment (marto83) [reply src/IO.Ably.Shared/Types/MessageExtras.cs]: Have a look at the latest commit. I update the class to be immutable.

File: src/IO.Ably.Shared/CustomSerialisers/MessageExtrasConverter.cs (hunk 2 above)
Comment (marto83) [reply src/IO.Ably.Shared/CustomSerialisers/MessageExtrasConverter.cs:10]: Done.

Comment (QuintinWillison) [reply src/IO.Ably.Shared/CustomSerialisers/MessageExtrasConverter.cs:10]: https://github.com/ably/ably-dotnet/pull/424/commits/7cbc8c72e64a4755fef43127ea386d8e33447e43

File: src/IO.Ably.Shared/Types/MessageExtras.cs (hunk 1 above)
Comment (QuintinWillison) [reply src/IO.Ably.Shared/Types/MessageExtras.cs]: https://github.com/ably/ably-dotnet/pull/424/commits/d14432ea6458a1d1d5bea5662292de0c0992df9c

File: src/IO.Ably.Shared/Types/MessageExtras.cs [hunk 3]
```diff
@@ -1,14 +1,64 @@
...
+        public DeltaExtras Delta { get; private set; }
+
+        /// <summary>
+        /// Messages extras is a flexible object that may other properties that are not exposed by the strongly typed implementation.
+        /// </summary>
+        /// <param name="data">the json object passed to Message extras.</param>
+        public MessageExtras(JToken data = null)
+        {
+            Data = data;
+            ParseData();
```
[the remark this answers is not in the document]
Comment (marto83) [reply src/IO.Ably.Shared/Types/MessageExtras.cs]: Good point. I'm sure there was a reason why I created the method in the beginning but I can no longer see it either. I moved the logic to the constructor and updated the `Delta` property to be readonly once the constructor has finished. I'm going to merge it to master

[the remark this answers is not in the document]
Comment (QuintinWillison) [reply src/IO.Ably.Shared/Types/MessageExtras.cs]: https://github.com/ably/ably-dotnet/pull/424/commits/c4ca8b8bf2fccb9283618950d672adff71f2320e

**Comment by marto83** [comment]
Merging the branch as the test failures are unrelated to the changes.

**Pull request merged**